File: NoTomorrow/Features/Dashboard/LastSessionRow.swift
import SwiftUI

/// "Last session" header (day · name · duration · n PRs) with one chip per PR set. Tapping opens that workout's
/// detail (the Train tab when there is none yet).
struct LastSessionRow: View {
    var workout: Workout?
    var units: WeightUnit = .kg
    var action: () -> Void

    var body: some View {
        Button(action: action) {
            VStack(alignment: .leading, spacing: 12) {
                HStack(alignment: .firstTextBaseline) {
                    Text("dashboard.lastSession").font(NT.Fonts.headline).foregroundStyle(NT.Colors.ink)
                    Spacer(minLength: 12)
                    if let workout { summary(for: workout) }
                }
                if let workout {
                    let prSets = prSets(in: workout)
                    if !prSets.isEmpty {
                        FlowChips(items: prSets.map { chipText($0) })
                    }
                } else {
                    Text("dashboard.noSessionsYet")
                        .font(NT.Fonts.subheadline)
                        .foregroundStyle(NT.Colors.ink2)
                }
            }
            .contentShape(Rectangle())
        }
        .buttonStyle(PressScale())
    }

    private func summary(for workout: Workout) -> some View {
        var text = Text(verbatim: "\(dayLabel(workout)) · \(workout.name) · \(Fmt.duration(workout.duration))")
        if workout.prCount > 0 {
            text = text + Text(verbatim: " · ")
                + Text(verbatim: String(format: String(localized: "dashboard.prs"), workout.prCount))
                    .foregroundStyle(NT.Colors.ember)
        }
        return text
            .font(NT.Fonts.subheadline)
            .foregroundStyle(NT.Colors.ink2)
            .tabular()
            .multilineTextAlignment(.trailing)
    }

    /// Weekday for anything in the last six days, otherwise day + month.
    private func dayLabel(_ workout: Workout) -> String {
        let when = workout.endedAt ?? workout.startedAt
        let cal = Calendar.current
        let weekAgo = cal.date(byAdding: .day, value: -6, to: cal.startOfDay(for: .now)) ?? .distantPast
        return when >= weekAgo ? Fmt.weekdayShort(when) : Fmt.dayMonth(when)
    }

    private func prSets(in workout: Workout) -> [SetEntry] {
        workout.sortedExercises.flatMap { $0.sortedSets.filter(\.isPR) }
    }

    private func chipText(_ set: SetEntry) -> String {
        let name = set.workoutExercise?.exercise?.localizedName ?? ""
        return "\(name) \(Fmt.set(set.weightKg, set.reps, unit: units))"
    }
}

/// Wrapping row of 32 pt trophy chips.
private struct FlowChips: View {
    var items: [String]

    var body: some View {
        FlowLayout(spacing: 8) {
            ForEach(Array(items.enumerated()), id: \.offset) { _, text in
                HStack(spacing: 6) {
                    Image(systemName: "trophy.fill")
                        .font(.system(size: 12, weight: .semibold))
                        .foregroundStyle(NT.Colors.ember)
                    Text(text)
                        .font(NT.Fonts.footnote)
                        .foregroundStyle(NT.Colors.ink)
                        .tabular()
                        .lineLimit(1)
                }
                .padding(.horizontal, 12)
                .frame(height: 32)
                .background(NT.Colors.surface, in: Capsule())
            }
        }
    }
}

/// Minimal left-to-right wrapping layout (iOS 16+ `Layout`).
private struct FlowLayout: Layout {
    var spacing: CGFloat = 8

    func sizeThatFits(proposal: ProposedViewSize, subviews: Subviews, cache: inout ()) -> CGSize {
        let width = proposal.width ?? .infinity
        return arrange(width: width, subviews: subviews).size
    }

    func placeSubviews(in bounds: CGRect, proposal: ProposedViewSize, subviews: Subviews, cache: inout ()) {
        let result = arrange(width: bounds.width, subviews: subviews)
        for (index, origin) in result.origins.enumerated() {
            subviews[index].place(at: CGPoint(x: bounds.minX + origin.x, y: bounds.minY + origin.y),
                                  proposal: .unspecified)
        }
    }

    private func arrange(width: CGFloat, subviews: Subviews) -> (size: CGSize, origins: [CGPoint]) {
        var origins: [CGPoint] = []
        var x: CGFloat = 0, y: CGFloat = 0, rowHeight: CGFloat = 0, maxX: CGFloat = 0
        for subview in subviews {
            let size = subview.sizeThatFits(.unspecified)
            if x > 0, x + size.width > width {
                x = 0
                y += rowHeight + spacing
                rowHeight = 0
            }
            origins.append(CGPoint(x: x, y: y))
            x += size.width + spacing
            rowHeight = max(rowHeight, size.height)
            maxX = max(maxX, x - spacing)
        }
        return (CGSize(width: maxX, height: y + rowHeight), origins)
    }
}

File: android/app/src/main/java/app/notomorrow/feature/dashboard/LastSessionRow.kt
package app.notomorrow.feature.dashboard

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.text.BasicText
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.SpanStyle
import androidx.compose.ui.text.buildAnnotatedString
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.text.withStyle
import androidx.compose.ui.unit.dp
import app.notomorrow.designsystem.NT
import app.notomorrow.designsystem.NtFlowLayout
import app.notomorrow.designsystem.NtIcon
import app.notomorrow.designsystem.NtIcons
import app.notomorrow.designsystem.NtShapes
import app.notomorrow.designsystem.NtText
import app.notomorrow.designsystem.TabularText
import app.notomorrow.designsystem.pressScale
import app.notomorrow.designsystem.sfIconSize
import app.notomorrow.designsystem.tabular
import app.notomorrow.model.WeightUnit
import app.notomorrow.util.Fmt
import app.notomorrow.util.S
import app.notomorrow.util.rememberNtStrings
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId

/**
 * "Last session" header (day · name · duration · n PRs) with one chip per PR set — the port
 * of `LastSessionRow` (`Features/Dashboard/LastSessionRow.swift`). Tapping opens that workout's
 * detail over Today (the Train tab when there is none yet).
 */
@Composable
fun LastSessionRow(
    session: LastSession?,
    units: WeightUnit,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
) {
    Column(
        modifier = modifier.fillMaxWidth().pressScale(onClick = onClick),
        verticalArrangement = Arrangement.spacedBy(12.dp),
    ) {
        Row(Modifier.fillMaxWidth()) {
            NtText(
                text = stringResource(S.dashboard_lastSession),
                modifier = Modifier.alignByBaseline(),
                style = NT.Fonts.headline,
                color = NT.Colors.ink,
                maxLines = 1,
            )
            Spacer(Modifier.weight(1f).widthIn(min = 12.dp))
            if (session != null) Summary(session, Modifier.alignByBaseline())
        }
        if (session != null) {
            if (session.prSets.isNotEmpty()) PrChips(session.prSets, units)
        } else {
            NtText(
                text = stringResource(S.dashboard_noSessionsYet),
                style = NT.Fonts.subheadline,
                color = NT.Colors.ink2,
            )
        }
    }
}

/** `"Fri · Legs · 52 min"`, plus `" · 2 PRs"` in ember when the session set records. */
@Composable
private fun Summary(session: LastSession, modifier: Modifier = Modifier) {
    val strings = rememberNtStrings()
    val name = app.notomorrow.feature.workout.workoutDisplayName(session.name)
    val text = buildAnnotatedString {
        append(dayLabel(session.at))
        append(" · ")
        append(name)
        append(" · ")
        append(Fmt.duration(session.durationSeconds, strings))
        if (session.prCount > 0) {
            append(" · ")
            withStyle(SpanStyle(color = NT.Colors.ember)) {
                append(stringResource(S.dashboard_prs, session.prCount))
            }
        }
    }
    BasicText(
        text = text,
        modifier = modifier,
        style = NT.Fonts.subheadline.tabular().copy(
            color = NT.Colors.ink2,
            textAlign = TextAlign.End,
        ),
        overflow = TextOverflow.Ellipsis,
    )
}

/** Weekday for anything in the last six days, otherwise day + month. */
private fun dayLabel(at: Instant, zone: ZoneId = ZoneId.systemDefault()): String {
    val date = at.atZone(zone).toLocalDate()
    val weekAgo = LocalDate.now(zone).minusDays(6)
    return if (!date.isBefore(weekAgo)) Fmt.weekdayShort(date) else Fmt.dayMonth(date)
}

/** Wrapping row of 32 dp trophy chips. */
@Composable
private fun PrChips(sets: List<PrSet>, units: WeightUnit) {
    NtFlowLayout(Modifier.fillMaxWidth(), spacing = 8.dp) {
        for (set in sets) {
            Row(
                modifier = Modifier
                    .height(32.dp)
                    .background(NT.Colors.surface, NtShapes.capsule)
                    .padding(horizontal = 12.dp),
                horizontalArrangement = Arrangement.spacedBy(6.dp),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                NtIcon(NtIcons.TrophyFill, size = sfIconSize(12f), tint = NT.Colors.ember)
                TabularText(
                    text = set.exerciseName + " " + Fmt.set(set.weightKg, set.reps, units),
                    style = NT.Fonts.footnote,
                    color = NT.Colors.ink,
                )
            }
        }
    }
}
